### PR TITLE
Add E0609

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #42229.

cc @Susurrus

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -4095,6 +4095,33 @@ assert_eq!(!Question::No, true);
 ```
 "##,
 
+E0609: r##"
+An attempt to get a field which doesn't exist was performed.
```
Comment (Susurrus) [inline src/librustc_typeck/diagnostics.rs]: Maybe "Attempted to access a non-existent field in a struct" reads better?

File: src/librustc_typeck/diagnostics.rs [hunk 2]
```diff
@@ -4095,6 +4095,33 @@ assert_eq!(!Question::No, true);
...
+```compile_fail,E0609
+struct StructWithFields {
+    x: u32,
+}
+
+let s = StructWithFields { x: 0 };
+println!("{}", s.foo); // error: no field `foo` on type `StructWithFields`
+```
+
+To fix this error, check that if you didn't mispell the field's name or that
```
Comment (Susurrus) [inline src/librustc_typeck/diagnostics.rs]: "misspell"

Remove the "if" as well.

**Review by Susurrus — CHANGES_REQUESTED** [review]

File: src/librustc_typeck/diagnostics.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: What about "An attempt to access a non-existent field in a struct"?

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by Susurrus** [comment]
I really don't like the phrasing "An attempt ... was performed". It reads awkwardly. I think using "Attempted to ..." reads better overall.

Additionally it appears that this error covers two situations: One where they improperly use field syntax on a type (currently non-structs-or-tuples) and another error when the struct does not contain the given field. And my understanding is that error is separate from when that field is private as well. I think these errors would be more clear if E0609 was split into two errors: one for when it's invalid for the type to have field access performed on it and another for when a struct or tuple doesn't have the field being referenced.

**Comment by GuillaumeGomez** [comment]
I splitted the error into two.

File: src/librustc_typeck/diagnostics.rs [hunk 3]
```diff
@@ -4095,6 +4095,63 @@ assert_eq!(!Question::No, true);
...
+```compile_fail,E0609
+struct StructWithFields {
+    x: u32,
+}
+
+let s = StructWithFields { x: 0 };
+println!("{}", s.foo); // error: no field `foo` on type `StructWithFields`
+```
+
+To fix this error, check if you didn't misspell the field's name or that the
```
Comment (Susurrus) [inline src/librustc_typeck/diagnostics.rs]: "check **that** you didn't misspell the field's name or that the field actually exist**s**"

File: src/librustc_typeck/diagnostics.rs [hunk 4]
```diff
@@ -4095,6 +4095,63 @@ assert_eq!(!Question::No, true);
...
+"##,
+
+E0610: r##"
+Attempted to access a field on a primitive type.
+
+Erroneous code example:
+
+```compile_fail,E0610
+let x: u32 = 0;
+println!("{}", x.foo); // error: `{interger}` is a primitive type, therefore
```
Comment (Susurrus) [inline src/librustc_typeck/diagnostics.rs]: "integer"

File: src/librustc_typeck/diagnostics.rs [hunk 5]
```diff
@@ -4095,6 +4095,63 @@ assert_eq!(!Question::No, true);
...
+                       //        doesn't have fields
+```
+
+Primitive types are the most basic types available in Rust and don't have
+fields. If you want to use fields, take a look to the structs. Example:
```
Comment (Susurrus) [inline src/librustc_typeck/diagnostics.rs]: I don't know how the user would accidentally access a field of a variable when coding. I would think the most likely reason would be they a) didn't know the datatype of the variable they used or b) accidentally used the wrong variable. I don't know if suggesting the user looks at structs will be helpful in them resolving this error. I think this extra description may be unnecessary.

In case you want to keep it, I'd reword the second sentence as it reads as being very informal. Maybe instead "To access data via named fields, struct types are used."?

File: src/librustc_typeck/diagnostics.rs [hunk 6]
```diff
@@ -4095,6 +4095,63 @@ assert_eq!(!Question::No, true);
...
+```
+// We declare struct called `Foo` containing two fields:
+struct Foo {
+    x: u32,
+    y: i64,
+}
+
+// We create an instance of this struct:
+let variable = Foo { x: 0, y: -12 };
+// And we can now call its fields:
```
Comment (Susurrus) [inline src/librustc_typeck/diagnostics.rs]: "call" is usually reserved for use with functions/methods. I'd suggest "access" instead.

File: src/librustc_typeck/diagnostics.rs [hunk 7]
```diff
@@ -4095,6 +4095,63 @@ assert_eq!(!Question::No, true);
...
+println!("x: {}, y: {}", variable.x, variable.y);
+```
+
+For more information, go read The Rust Book: https://doc.rust-lang.org/book/
```
Comment (Susurrus) [inline src/librustc_typeck/diagnostics.rs]: "go read" sounds like a derogatory command. I'd suggest "see" instead (and drop the comma).

File: src/librustc_typeck/check/mod.rs [hunk 8]
```diff
@@ -2921,30 +2921,34 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                    }
+                    _ => {}
                 }
-                _ => {}
-            }
-            err.emit();
+                err
+            } else {
+                type_error_struct!(self.tcx().sess, field.span, expr_t, E0610,
+                                   "`{}` is a primitive type, therefore doesn't have fields",
```
Comment (Susurrus) [inline src/librustc_typeck/check/mod.rs]: I'd suggest replacing the comma with "and"

**Review by Susurrus — CHANGES_REQUESTED** [review]

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/librustc_typeck/diagnostics.rs [hunk 9]
```diff
@@ -4122,6 +4122,36 @@ println!("{}", s.x); // ok!
...
+    y: i64,
+}
+
+// We create an instance of this struct:
+let variable = Foo { x: 0, y: -12 };
+// And we can now access its fields:
+println!("x: {}, y: {}", variable.x, variable.y);
+```
+
+For more information, see The Rust Book: https://doc.rust-lang.org/book/
```
Comment (Susurrus) [inline src/librustc_typeck/diagnostics.rs]: Remove the comma as it's unnecessary.

File: src/librustc_typeck/diagnostics.rs [hunk 10]
```diff
@@ -4096,7 +4096,7 @@ assert_eq!(!Question::No, true);
 "##,
 
 E0609: r##"
-An attempt to access a non-existent field in a struct was performed.
+Attempted to access a non-existent field in a struct was performed.
```
Comment (Susurrus) [inline src/librustc_typeck/diagnostics.rs]: Remove "was performed", the sentence doesn't make sense given that it now starts with "Attempted"

**Review by Susurrus — CHANGES_REQUESTED** [review]

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by Susurrus** [comment]
LGTM 👍
